### PR TITLE
Refreshing Cache

### PR DESCRIPTION
**Pull request by kailuowang**
Submitting this PR to solicit some feedback 
It's very different design from the other `AutoFetchingCache`. 

- It's a super set of the MemoryCache built on top it. 
- Each item can optionally have its own refresh cycle and fetch effect
- Failing refreshes eventually will let the value expire (for now it always use the default expiration time, will add custom expiration later)
- I didn't provide a `BoundedRefresh`. Personally I feel there is a lot of complexity involved, and we need a lot of tests to cover them. 
- ~Right now like `AutoFetchCache`, errors during fetch are just swollen, but I think we can allow custom error handlers: PartialFunction[Throwable, Unit]~  updated I added item specific recovery
- I didn't use cats-collection which could be a reason that we provide it in a different module.

### CODE REVIEW

File: modules/reload/src/main/scala/io/chrisdavenport/mules/reload/RefreshingCache.scala [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+          case None => NotFound.pure[F].widen
+        }
+      }
+
+      for {
+        fiber <- Concurrent[F].start(
+          Timer[F].sleep(Duration.fromNanos(period.nanos)) >>
+            doRefreshFetch
+        )
+        result <- fiber.join
```
Comment (ChristopherDavenport) [inline modules/reload/src/main/scala/io/chrisdavenport/mules/reload/RefreshingCache.scala]: Why do we immediately join after starting?

Comment (kailuowang) [reply modules/reload/src/main/scala/io/chrisdavenport/mules/reload/RefreshingCache.scala]: My bad, I kind of inherited this code from [AutoFetchingCache](https://github.com/ChristopherDavenport/mules/blob/master/modules/reload/src/main/scala/io/chrisdavenport/mules/reload/AutoFetchingCache.scala#L114-L119) 
Now you pointed it out I don't know if it's needed there. I will change this. 
Another thing I noticed in `AutoFetchingCache` is that all concurrency reads are going through guarded by the Semaphore, which might be a bottleneck.

**Comment by kailuowang** [comment]
The main caveat is that this cache is unbounded. Auto refreshing items add to the cache will live forever. So it will not be safe for items to be added programmatically, which the API allows to do. 

I was thinking about having a usage time out, i.e. if an item remain unused for over a period, it's automatically purged. But I couldn't think of a way to make concurrent reads blocked by a lock (since now every read needs to update a timestamp to mark the last usage)
So until we can resolve a auto purge mechanism,  I probably won't use it in my own projects. (I will use Mau instead). 

I updated it, but given the caveat above feel free to leave this unmerged or simply close it. I can reopen when i have a solution.

**Comment by kailuowang** [comment]
I am going to close this one for now. I can provide support if there is interest from other users. but for my own usage, [mau](https://github.com/kailuowang/mau) seems to be working well in production.

**Pull request closed without merging**